File: app/src/main/java/org/andbootmgr/app/legacy/util/SDUtils.java
package org.andbootmgr.app.legacy.util;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.topjohnwu.superuser.Shell;

import org.andbootmgr.app.R;
import org.andbootmgr.app.legacy.devices.DeviceModel;
import org.andbootmgr.app.util.SOUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.stream.Collectors;

public class SDUtils {

    public static final HashMap<String, String> codes = new HashMap<>();

    public static void setupCodes(Context c) {
        codes.put("0700", c.getString(R.string.portable_part));
        codes.put("8302", c.getString(R.string.data_part));
        codes.put("8301", c.getString(R.string.meta_part));
        codes.put("8305", c.getString(R.string.system_part));
        codes.put("8300", c.getString(R.string.unknown_part));
    }

    public enum PartitionType {
        RESERVED, ADOPTED, PORTABLE, UNKNOWN, FREE, SYSTEM, DATA
    }

    public static class Partition {
        public PartitionType type;
        public int id;
        public long startSector;
        public long endSector;
        public long size;
        public String sizeFancy;
        public String code;
        public String name;
        public int minor;

        @NonNull
        @Override
        public String toString() {
            return "Partition{" +
                    "type=" + type +
                    ", id=" + id +
                    ", startSector=" + startSector +
                    ", endSector=" + endSector +
                    ", size=" + size +
                    ", sizeFancy='" + sizeFancy + '\'' +
                    ", code='" + code + '\'' +
                    ", name='" + name + '\'' +
                    ", minor=" + minor +
                    '}';
        }

        public static class FreeSpace extends Partition {

            public FreeSpace(long start, long end, int bytes) {
                startSector = ((start / 2048) + 1) * 2048 ; endSector = end; size = end - start; type = PartitionType.FREE; sizeFancy = SOUtils.humanReadableByteCountBin(size * bytes);
            }

            @NonNull
            @Override
            public String toString() {
                return "FreeSpace{" +
                        "startSector=" + startSector +
                        ", endSector=" + endSector +
                        ", size=" + size +
                        '}';
            }
        }
    }

    public static class SDPartitionMeta {
        @SuppressWarnings("CanBeFinal")
        public List<Partition> p = new ArrayList<>();
        @SuppressWarnings("CanBeFinal")
        public List<Partition> u = new ArrayList<>();
        public List<Partition> s = new ArrayList<>();
        public String friendlySize;
        public String guid;
        public long sectors;
        public int logicalSectorSizeBytes;
        public int maxEntries;
        public long firstUsableSector;
        public long lastUsableSector;
        public long alignSector;
        public long totalFreeSectors;
        public String totalFreeFancy;
        public long usableSectors;
        public int nid;
        public int major;
        public int minor;
        public String path;
        public String ppath;

        @NonNull
        @Override
        public String toString() {
            return "SDPartitionMeta{" +
                    "p=" + p +
                    ", s=" + s +
                    ", friendlySize='" + friendlySize + '\'' +
                    ", guid='" + guid + '\'' +
                    ", sectors=" + sectors +
                    ", logicalSectorSizeBytes=" + logicalSectorSizeBytes +
                    ", maxEntries=" + maxEntries +
                    ", firstUsableSector=" + firstUsableSector +
                    ", lastUsableSector=" + lastUsableSector +
                    ", alignSector=" + alignSector +
                    ", totalFreeSectors=" + totalFreeSectors +
                    ", totalFreeFancy='" + totalFreeFancy + '\'' +
                    ", usableSectors=" + usableSectors +
                    '}';
        }

        public int countPartition() {
            return p.size();
        }

        public Partition dumpPartition(int id) {
            return p.get(id);
        }

        public int count() { return u.size(); }

        public Partition dump(int id) {
            return u.get(id);
        }

        public Partition dumpS(int id) { return s.get(id); }
    }

    public static String umsd(PartitionType t, int major, int minor) {
        switch (t) {
            case FREE:
                return "true";
            case PORTABLE:
                return "sm unmount public:" + major + "," + minor;
            case ADOPTED:
                return "sm unmount private:" + major + "," + minor;
            case RESERVED:
            case UNKNOWN:
            default:
                return "echo 'Warning: Unsure on how to unmount this partition.'";
        }
    }

    public static String umsd(SDPartitionMeta meta) {
        StringBuilder s = new StringBuilder();
        for (Partition p : meta.p)
            s.append(umsd(p.type, meta.major, p.minor)).append(" && ");
        String e = s.toString();
        if (e.isEmpty()) return "true";
        return e.substring(0, e.length()-4);
    }

    @Nullable
    public static SDPartitionMeta generateMeta(String bdev, String pbdev) {
        SDPartitionMeta meta;
        Shell.Result r = Shell.su("printf \"mm:%d:%d\\n\" `stat -c '0x%t 0x%T' " + bdev + "` && sgdisk " + bdev + " --print").exec();
        if (r.isSuccess())
            meta = new SDPartitionMeta();
        else
            return null;
        meta.path = bdev;
        meta.ppath = pbdev;
        meta.nid = 1;
        long temp = -1;
        for (String o : r.getOut()) {
            if (o.startsWith("Disk ") && !o.contains("GUID")) {
                String[] t = o.split(": ")[1].split(", ");
                meta.sectors = Long.parseLong(t[0].replace(" sectors",""));
                meta.friendlySize = t[1];
            } else if(o.startsWith("Logical sector size: ") && o.endsWith(" bytes")) {
                meta.logicalSectorSizeBytes = Integer.parseInt(o.replace("Logical sector size: ","").replace(" bytes",""));
            } else if(o.startsWith("Sector size (logical/physical): ") && o.endsWith(" bytes")) {
                meta.logicalSectorSizeBytes = Integer.parseInt(o.replace("Sector size (logical/physical): ","").replace(" bytes","").split("/")[0]);
            } else if(o.startsWith("Disk identifier (GUID): ")) {
                meta.guid = o.replace("Disk identifier (GUID): ", "");
            } else if (o.startsWith("Partition table holds up to ")) {
                meta.maxEntries = Integer.parseInt(o.replace("Partition table holds up to ","").replace(" entries",""));
            } else if (o.startsWith("First usable sector is ")) {
                meta.firstUsableSector = Long.parseLong(o.replace("First usable sector is ","").replaceFirst(", last usable sector is .*$",""));
                meta.lastUsableSector = Long.parseLong(o.replace("First usable sector is " + meta.firstUsableSector + ", last usable sector is ",""));
                meta.usableSectors = meta.lastUsableSector - meta.firstUsableSector;
                temp = meta.firstUsableSector;
            } else if (o.startsWith("Partitions will be aligned on ") && o.endsWith("-sector boundaries")) {
                meta.alignSector = Long.parseLong(o.replace("Partitions will be aligned on ","").replace("-sector boundaries",""));
            } else if (o.startsWith("Total free space is ") && o.contains("sectors")) {
                String[] t = o.replace(")", "").split("\\(");
                meta.totalFreeSectors = Long.parseLong(t[0].replace("Total free space is ", "").replace(" sectors ", ""));
                meta.totalFreeFancy = t[1];
            } else if(o.startsWith("mm:")) {
                String[] t = o.trim().split(":");
                meta.major = Integer.parseInt(t[1]);
                meta.minor = Integer.parseInt(t[2]);
            } else if(o.equals("") || o.startsWith("Number  Start (sector)    End (sector)  Size       Code  Name") || o.startsWith("Main partition table begins at")) {
                assert true; //avoid empty statement warning but do nothing
            } else if (o.startsWith("  ") && o.contains("iB")) {
                while (o.contains("  "))
                    o = o.trim().replace("  "," ");
                String[] ocut = o.split(" ");
                Partition p = new Partition();
                p.id = Integer.parseInt(ocut[0]);
                if (meta.nid == p.id) meta.nid++;
                p.minor = meta.minor + p.id;
                p.startSector = Long.parseLong(ocut[1]);
                p.endSector = Long.parseLong(ocut[2]);
                p.sizeFancy = ocut[3] + " " + ocut[4];
                p.size = p.endSector - p.startSector;
                p.code = ocut[5];
                p.name = String.join(" ", Arrays.copyOfRange(ocut, 6, ocut.length));
                switch (p.code) {
                    case "8301":
                        p.type = p.id == 1 ? PartitionType.RESERVED : PartitionType.UNKNOWN;
                        break;
                    case "0700":
                        p.type = PartitionType.PORTABLE;
                        break;
                    case "FFFF":
                        p.type = p.id == 1 ? PartitionType.RESERVED : PartitionType.ADOPTED;
                        break;
                    case "8302":
                        p.type = PartitionType.DATA;
                        break;
                    case "8305":
                        p.type = PartitionType.SYSTEM;
                        break;
                    case "8300":
                    default:
                        p.type = PartitionType.UNKNOWN;
                        break;
                }
                meta.p.add(p);
                meta.u.add(p);
            } else {
                Log.e("ABM SDUtils","can't handle " + o);
                return null;
            }
        }
        @SuppressWarnings({"FuseStreamOperations", "SimplifyStreamApiCallChains"}) List<Partition> l = meta.u.stream().collect(Collectors.toList()); // this actually copies, therefore, we need this.
        l.sort((o1, o2) -> {
            // this is like this because startSector is long and I don't want overflows due to casting
            if(o1.startSector - o2.startSector < -1)
                return -1;
            else if (o1.startSector == o2.startSector)
                return 0;
            else return 1;
        });
        for (Partition p : l) {
            if (p.startSector > temp + meta.alignSector)
                meta.u.add(new Partition.FreeSpace(temp, p.startSector-1, meta.logicalSectorSizeBytes));
            temp = p.endSector;
        }
        if (meta.lastUsableSector > temp + meta.alignSector)
            meta.u.add(new Partition.FreeSpace(temp, meta.lastUsableSector-1, meta.logicalSectorSizeBytes));
        meta.s = meta.u.subList(0, meta.u.size());
        meta.s.sort((o1, o2) -> {
            // this is like this because startSector is long and I don't want overflows due to casting
            if(o1.startSector - o2.startSector < -1)
                return -1;
            else if (o1.startSector == o2.startSector)
                return 0;
            else return 1;
        });
        Log.i("ABM",meta.toString());
        return meta;
    }

    public static SDPartitionMeta generateMeta(DeviceModel m) {
        return generateMeta(m.bdev, m.pbdev);
    }

}

File: app/src/main/java/org/andbootmgr/app/util/SDUtils.kt
package org.andbootmgr.app.util

import android.util.Log
import com.topjohnwu.superuser.Shell
import org.andbootmgr.app.DeviceInfo
import java.util.*
import java.util.stream.Collectors
import kotlin.jvm.optionals.getOrElse

object SDUtils {

	// Unmount drive
	fun umsd(meta: SDPartitionMeta): String {
		val s = StringBuilder()
		for (p in meta.p) s.append(p.unmount()).append(" && ")
		val e = s.toString()
		return if (e.isEmpty()) "true" else e.substring(0, e.length - 4)
	}

	fun generateMeta(deviceInfo: DeviceInfo): SDPartitionMeta? {
		val meta: SDPartitionMeta
		val r =
			Shell.cmd("printf \"mm:%d:%d\\n\" `stat -c '0x%t 0x%T' ${deviceInfo.bdev}` && sgdisk ${deviceInfo.bdev} --print")
				.exec()
		meta = if (r.isSuccess) SDPartitionMeta() else return null
		try {
			meta.path = deviceInfo.bdev
			meta.ppath = deviceInfo.pbdev
			meta.nid = 1
			var temp: Long = -1
			var o: String
			for (oo in r.out) {
				o = oo
				if (o.startsWith("Disk ") && !o.contains("GUID")) {
					val t = o.split(": ").toTypedArray()[1].split(", ").toTypedArray()
					meta.sectors = t[0].replace(" sectors", "").toLong()
					meta.friendlySize = t[1]
				} else if (o.startsWith("Logical sector size: ") && o.endsWith(" bytes")) {
					meta.logicalSectorSizeBytes =
						o.replace("Logical sector size: ", "").replace(" bytes", "").toInt()
				} else if (o.startsWith("Sector size (logical/physical): ") && o.endsWith(" bytes")) {
					meta.logicalSectorSizeBytes =
						o.replace("Sector size (logical/physical): ", "").replace(" bytes", "")
							.split("/").toTypedArray()[0].toInt()
				} else if (o.startsWith("Disk identifier (GUID): ")) {
					meta.guid = o.replace("Disk identifier (GUID): ", "")
				} else if (o.startsWith("Partition table holds up to ")) {
					meta.maxEntries =
						o.replace("Partition table holds up to ", "").replace(" entries", "").toInt()
				} else if (o.startsWith("First usable sector is ")) {
					meta.firstUsableSector = o.replace("First usable sector is ", "")
						.replaceFirst(", last usable sector is .*$".toRegex(), "").toLong()
					meta.lastUsableSector = o.replace(
						"First usable sector is " + meta.firstUsableSector + ", last usable sector is ",
						""
					).toLong()
					meta.usableSectors = meta.lastUsableSector - meta.firstUsableSector
					temp = meta.firstUsableSector
				} else if (o.startsWith("Partitions will be aligned on ") && o.endsWith("-sector boundaries")) {
					meta.alignSector = o.replace("Partitions will be aligned on ", "")
						.replace("-sector boundaries", "").toLong()
				} else if (o.startsWith("Total free space is ") && o.contains("sectors")) {
					val t = o.replace(")", "").split("(").toTypedArray()
					meta.totalFreeSectors =
						t[0].replace("Total free space is ", "").replace(" sectors ", "").toLong()
					meta.totalFreeFancy = t[1]
				} else if (o.startsWith("mm:")) {
					val t = o.trim().split(":").toTypedArray()
					meta.major = t[1].toInt()
					meta.minor = t[2].toInt()
				} else if (o == "" || o.startsWith("Number  Start (sector)    End (sector)  Size       Code  Name") || o.startsWith(
						"Main partition table begins at"
					)
				) {
					assert(true) //avoid empty statement warning but do nothing
				} else if (o.startsWith("  ") && o.contains("iB")) {
					while (o.contains("  ")) o = o.trim { it <= ' ' }.replace("  ", " ")
					val ocut = o.split(" ").toTypedArray()
					val id = ocut[0].toInt()
					val code = ocut[5]
					val type = when (code) {
						"8301" ->
							if (id == 1) PartitionType.RESERVED else PartitionType.UNKNOWN
						"0700" -> PartitionType.PORTABLE
						"FFFF" ->
							if (id == 1) PartitionType.RESERVED else PartitionType.ADOPTED
						"8302" -> PartitionType.DATA
						"8305" -> PartitionType.SYSTEM
						"8300" -> PartitionType.UNKNOWN
						else -> PartitionType.UNKNOWN
					}
					val p = Partition(
						meta,
						type,
						id,
						meta.ppath + id,
						ocut[1].toLong() /* startSector */,
						ocut[2].toLong()/* endSector */,
						meta.logicalSectorSizeBytes,
						code,
						ocut.copyOfRange(6, ocut.size).toList().joinToString(" ") /* label/name */,
						meta.major,
						meta.minor + id
					)
					if (meta.nid == p.id) meta.nid++
					meta.p.add(p)
					meta.u.add(p)
				} else if (o=="Found invalid GPT and valid MBR; converting MBR to GPT format"){
					meta.ismbr=true
				} else if(o=="***************************************************************" || o=="in memory. ") {
					assert(true)
				} else {
					Log.e("ABM SDUtils", "can't handle $o")
					return null
				}
			}
			val l = meta.u.stream()
				.collect(Collectors.toList()) // this actually copies, therefore, we need this.
			l.sortWith { o1: Partition, o2: Partition ->
				// this is like this because startSector is long and I don't want overflows due to casting
				if (o1.startSector - o2.startSector < -1) return@sortWith -1 else if (o1.startSector == o2.startSector) return@sortWith 0 else return@sortWith 1
			}
			for (p in l) {
				if (p.startSector > temp + meta.alignSector) meta.u.add(
					Partition.FreeSpace(
						meta,
						temp,
						p.startSector - 1,
						meta.logicalSectorSizeBytes
					)
				)
				temp = p.endSector
			}
			if (meta.lastUsableSector > temp + meta.alignSector) meta.u.add(
				Partition.FreeSpace(
					meta,
					temp,
					meta.lastUsableSector - 1,
					meta.logicalSectorSizeBytes
				)
			)
			meta.s = meta.u.subList(0, meta.u.size)
			meta.s.sortWith { o1: Partition, o2: Partition ->
				// this is like this because startSector is long and I don't want overflows due to casting
				if (o1.startSector - o2.startSector < -1) return@sortWith -1 else if (o1.startSector == o2.startSector) return@sortWith 0 else return@sortWith 1
			}
			Log.i("ABM", meta.toString())
			return meta
		} catch (e: Exception) {
			Log.e("ABM", Log.getStackTraceString(e))
			return null
		}
	}

	enum class PartitionType {
		RESERVED, ADOPTED, PORTABLE, UNKNOWN, FREE, SYSTEM, DATA
	}

	open class Partition(
		var meta: SDPartitionMeta,
		val type: PartitionType,
		val id: Int,
		open val path: String,
		val startSector: Long,
		val endSector: Long,
		bytes: Int,
		val code: String,
		val name: String,
		val major: Int,
		val minor: Int
	) {
		val size: Long = endSector - startSector
		val sizeFancy: String = SOUtils.humanReadableByteCountBin(size * bytes)
		override fun toString(): String {
			return "Partition{" +
					"type=" + type +
					", id=" + id +
					", startSector=" + startSector +
					", endSector=" + endSector +
					", size=" + size +
					", sizeFancy='" + sizeFancy + '\'' +
					", code='" + code + '\'' +
					", name='" + name + '\'' +
					", minor=" + minor +
					'}'
		}

		fun mount(): String {
			return when (this.type) {
				PartitionType.FREE -> "echo 'Why are you trying to mount free space?!'"
				PartitionType.PORTABLE -> "sm mount public:${this.major},${this.minor}"
				PartitionType.ADOPTED -> "sm mount private:${this.major},${this.minor}"
				else -> "echo 'Warning: Unsure on how to mount this partition.'"
			}
		}
		fun unmount(): String {
			return when (type) {
				PartitionType.FREE -> "true"
				PartitionType.PORTABLE -> "sm unmount public:${this.major},${this.minor}"
				PartitionType.ADOPTED -> "sm unmount private:${this.major},${this.minor}"
				PartitionType.RESERVED -> {
					if (name == "abm_settings") {
						// TODO unmount bootset
						"true"
					} else {
						"echo 'Warning: Unsure on how to unmount this partition.'"
					}
				}
				PartitionType.SYSTEM, PartitionType.DATA -> {
					// TODO rework this when dual android is supported by looking at current os' xpart
					"true"
				}
				else -> "echo 'Warning: Unsure on how to unmount this partition.'"
			}
		}

		fun delete(): String {
			if (type == PartitionType.FREE)
				return "echo 'Tried to delete free space'"
			return "sgdisk " + meta.path + " --delete " + id
		}
		fun rename(newName: String): String {
			if (type == PartitionType.FREE)
				return "echo 'Tried to rename free space'; exit 1"
			return "sgdisk " + meta.path + " --change-name " + id + ":'" + newName.replace("'", "") + "'"
		}

		class FreeSpace(meta: SDPartitionMeta, start: Long, end: Long, bytes: Int) :
			Partition(meta,
				PartitionType.FREE, 0, "", (start / 2048 + 1) * 2048, end, bytes, "", "", 0, 0) {

			override val path
				get() = throw IllegalArgumentException()

			override fun toString(): String {
				return "FreeSpace{" +
						"startSector=" + startSector +
						", endSector=" + endSector +
						", size=" + size +
						", sizeFancy='" + sizeFancy + '\'' +
						'}'
			}

			// start & end are RELATIVE to startSector of this instance
			fun create(start: Long, end: Long, typecode: String, name: String): String {
				val rstart = startSector + start
				val rend = startSector + end
				val a = rstart > endSector
				val b = start < 0
				val c = rend > endSector
				val d = end < start
				if (a || b || c || d) {
					return "echo 'Invalid values ($start:$end - $rstart>$startSector:$endSector>$rend - $a $b $c $d). Aborting...'; exit 1"
				}
				return "sgdisk ${meta.path} --new ${meta.nid}:$rstart:$rend --typecode ${meta.nid}:$typecode --change-name ${meta.nid}:'${name.replace("'", "")}' && sleep 1 && ls ${meta.ppath}${meta.nid}" + when(typecode) {
					 "0700" -> " && sm format public:${meta.major},${meta.minor+meta.nid}"
					 "8301" -> " && mkfs.ext4 ${meta.ppath}${meta.nid}"
					 else -> ""
				}
			}
		}
	}

	class SDPartitionMeta {
		// List of partitions sorted by partition number
		var p: MutableList<Partition> = ArrayList()
		// List of partitions with free space entries, unsorted
		var u: MutableList<Partition> = ArrayList()
		// List of partitions with free space entries, sorted by physical order
		var s: MutableList<Partition> = ArrayList()
		var friendlySize: String? = null
		var guid: String? = null
		var sectors: Long = 0
		var logicalSectorSizeBytes = 0
		var maxEntries = 0
		var firstUsableSector: Long = 0
		var lastUsableSector: Long = 0
		var ismbr : Boolean = false
		var alignSector: Long = 0
		var totalFreeSectors: Long = 0
		var totalFreeFancy: String? = null
		var usableSectors: Long = 0
		// First free partition number
		var nid = 0
		var major = 0
		var minor = 0
		var path: String? = null
		var ppath: String? = null
		override fun toString(): String {
			return "SDPartitionMeta{" +
					"p=" + p +
					", s=" + s +
					", friendlySize='" + friendlySize + '\'' +
					", guid='" + guid + '\'' +
					", sectors=" + sectors +
					", logicalSectorSizeBytes=" + logicalSectorSizeBytes +
					", maxEntries=" + maxEntries +
					", firstUsableSector=" + firstUsableSector +
					", lastUsableSector=" + lastUsableSector +
					", alignSector=" + alignSector +
					", totalFreeSectors=" + totalFreeSectors +
					", totalFreeFancy='" + totalFreeFancy + '\'' +
					", usableSectors=" + usableSectors +
					'}'
		}

		// Get partition by kernel id
		fun dumpKernelPartition(id: Int): Partition {
			return p.stream().filter { it.id == id }.findFirst().getOrElse {
				throw IllegalArgumentException("no such partition with id $id, have: ${this.p}") }
		}
	}
}
